### DISCUSSION TITLE
Version 2.2.1

### DISCUSSION THREAD

**Discussion by saran-t** [google-deepmind/mujoco#387] [Announcements]
## General

- Added `mjd_transitionFD` to compute efficient finite difference approximations of the state-transition and
  control-transition matrices, [see here](https://mujoco.readthedocs.io/en/2.2.1/computation.html#derivatives) for more details.
- Added derivatives for the ellipsoid fluid model.
- Added `ctrl` attribute to [keyframes](https://mujoco.readthedocs.io/en/2.2.1/XMLreference.html#keyframe).
- Added `clock` sensor which [measures time](https://mujoco.readthedocs.io/en/2.2.1/XMLreference.html#sensor-clock).
- Added visualisation groups to skins.
- Added actuator visualisation for `free` and `ball` joints and for actuators with `site` transmission.
- Added visualisation for actuator activations.
- Added `<intvelocity>` actuator shortcut for "integrated velocity" actuators, documented [here](https://mujoco.readthedocs.io/en/2.2.1/XMLreference.html#intvelocity).
- Added `<damper>` actuator shortcut for active-damping actuators, documented [here](https://mujoco.readthedocs.io/en/2.2.1/XMLreference.html#damper).
- `mju_rotVecMat` and `mju_rotVecMatT` now support in-place multiplication.
- `mjData.ctrl` values are no longer clamped in-place, remain untouched by the engine.
- Arrays in mjData's buffer now align to 64-byte boundaries rather than 8-byte.
- Add memory poisoning when building with Address Sanitizer (ASAN) and Memory Sanitizer (MSAN). This allows ASAN to
  detect reads and writes to regions in `mjModel.buffer` and `mjData.buffer` that do not lie within an array, and
  for MSAN to detect reads from uninitialised fields in `mjData` following `mj_resetData`.
- Add a [slider-crank example](https://github.com/deepmind/mujoco/tree/2.2.1/model/slider_crank) to `model/`.

## Bug fixes

- [Activation clamping](https://mujoco.readthedocs.io/en/2.2.1/modeling.html#cactrange) was not being applied in the [implicit integrator](https://mujoco.readthedocs.io/en/2.2.1/computation.html#geintegration).
- Stricter parsing of orientation specifiers. Before this change, a specification that included both `quat` and an
  [alternative specifier](https://mujoco.readthedocs.io/en/2.2.1/modeling.html#corientation) e.g., `<geom ... quat=".1 .2 .3 .4" euler="10 20 30">`, would lead to the
  `quat` being ignored and only `euler` being used. After this change a parse error will be thrown.
- Stricter parsing of XML attributes. Before this change an erroneous XML snippet like `<geom size="1/2 3 4">` would
  have been parsed as `size="1 0 0"` and no error would have been thrown. Now throws an error.
- Trying to load a `NaN` via XML like `<geom size="1 NaN 4">`, while allowed for debugging purposes, will now print
  a warning.
- Fixed null pointer dereference in `mj_loadModel`.
- Fixed memory leaks when loading an invalid model from MJB.
- Integer overflows are now avoided when computing `mjModel` buffer sizes.
- Added missing warning string for `mjWARN_BADCTRL`.

## Packaging

- Changed MacOS packaging so that the copy of `mujoco.framework` embedded in `MuJoCo.app` can be used to build
  applications externally.

<hr /><em>This discussion was created from the release <a href='https://github.com/deepmind/mujoco/releases/tag/2.2.1'>2.2.1</a>.</em>

**Comment by kevinzakka** [comment]
Very exciting release, thanks @saran-t, @nimrod-gileadi, @yuvaltassa and team!!
